sdk/servicebus/service-bus/src/core/streamingReceiver.ts

```diff
@@ -94,6 +96,65 @@ export class StreamingReceiver extends MessageReceiver {
     }
   }
 
+  /**
+   * Unregisters this streaming receiver, removing it's callbacks
+   * and draining any credits that we requested.
+   *
+   * @param timeoutMs The maximum amount of time to wait for the drain to complete.
+   */
+  async drainAndUnregisterHandlers(timeoutMs: number): Promise<void> {
+    const receiver = this._receiver;
+
+    if (receiver == null) {
+      return;
+    }
+
+    log.receiver(
+      "[%s] Stopping streaming receiver for entity '%s'.",
+      this._context.namespace.connectionId,
+      this._context.entityPath
+    );
+
+    // drain this receiver - once it's fully drained
+    // we can kill all the other methods
+    const ret = await Promise.race([this._drainReceiver(receiver), delay(timeoutMs, "timedout")]);
+
+    if (ret === "timedout") {
+      const msg = `Failed to drain receiver within ${timeoutMs} milliseconds, can't safely close message handler.`;
+      log.error(`[${this._context.namespace.connectionId}] ${msg}`);
+      throw new Error(msg);
+    }
+
+    // succesfully drain()'d - we should receive no further messages
+    // or errors related this receiver so we can remove the handlers.
+    this._onMessage = async () => {};
+    this._onError = () => {};
```

Comment:
I see how unsetting the user provided handlers can result in user not receiving any more messages, but am not sure if this necessarily means that the library has stopped receiving messages from the service.

By the time we reach this line, we would have drained any credits on the receiver link, but that doesnt stop from more credits being added. Imagine the below sequence of events

- user set `maxConcurrentCalls` to 100, this results in us adding 100 credits on the line. After user call back is executed for a message, we add another credit
- say rhea received 10 messages, fires off events for all 10, there are 90 credits left
- user calls close, you drain the 90 credits
- user's callback for the 10 messages complete, resulting in us adding 10 credits on the line
- rhea receives 10 messages, but user never hears of this.

Reply:
Another sequence of events:
- same as above, but after user calls close and you drain the 90 credits, the connection goes down
- Our link recovery logic will bring the link back up and add the 100 credits right back on

Reply:
I'd like to recommend the below approach:
- For non sessions, user calling close() simply calls `streamingReceiver.close()`. 
     - Since we have the backup for settling messages, we are covered regarding messages that user is holding
     - Since the context cache is cleared, `disconnect` on the connection will not result in any recovery attempts
- For sessions, do the draining like you are doing here as well as add a check to add no more credits if the user has called close().
     - we cant close the link like we did for non sessions as we need it open to settle messages that the user is holding
     - since we don't recover receiver links for sessions ever in Track 1, we are good there
     - we need to remember to ensure to add credits as before when user calls the `registerMessageHandler` again after calling close() or throw an error that they need to create a new receiver

Reply:
Also, what is the guarantee that we want to provide the user here?
That their callback will not be called, but there is risk that messages will be lost (if in receive and delete mode) or their delivery count incremented?

Because even if you took care of the cases I mentioned in the previous comments, there is till the cases of messages received by rhea which are waiting in the event loop to call the user's callback.

Reply:
> Also, what is the guarantee that we want to provide the user here?
> That their callback will not be called, but there is risk that messages will be lost (if in receive and delete mode) or their delivery count incremented?
> 
> Because even if you took care of the cases I mentioned in the previous comments, there is till the cases of messages received by rhea which are waiting in the event loop to call the user's callback.

I think this is the critical one and should be the focus. It looks like I missed quite a bit - our abstractions could use some improvement!

It seems like I need:
- A way to ensure that there are no pending callbacks for the users processMessage (ie, I need to actually inspect this). 
- A way to prevent adding credits when we're trying to drain (ie, please stop making my job harder)
   - Reconnect needs to respect the same flag

Talking offline, I see why you want to branch the code, but let's see if we can avoid that so we don't have to bubble up all the session/non-session differences.

Reply:
> A way to ensure that there are no pending callbacks for the users processMessage

Unfortunately, I don't see a way to do this unless we are keeping track of each message for which our callback to rhea got fired and each message for which user's callback to us got fired.

> A way to prevent adding credits when we're trying to drain

Not just when you are trying to drain, but after the drain is completed too.